Thermocouple improvements

adq opened the pull request:
Hi, sorry about this deluge of patches from me, I've just been on a bit of a binge with this recently! Almost done I think/hope.

Anyway, this patch attempts to improve the thermocouple handling in various ways:

# Thermocouple errors at high temperature.

I have exactly the problem described here: https://forums.adafruit.com/viewtopic.php?f=31&t=169135#p827564

It seems to be a problem with the design of my kiln from the manufacturer. I'm wondering if its perhaps a difference between the **tiny** space in my jewellery kiln vs a ceramics kiln, but that is total speculation on my part.

When it gets above ~900C, I start getting erroneous error reports from the max31855 - most of the readings will claim there's an error, and it'll constantly flip between vcc_short and gnd_short at random (below 900 it is **totally** clean). However - as the link above says - the temperature reading still appears to be correct.

The original controller from my kiln works fine, and is handling the thermocouple using an opamp and a resistor/capacitor network: either its just ignoring the issue, or it has some circuitry not present on the max31855 that is somehow compensating. I believe it is an Orton Autofire controller.

Anyway, I can't really fix this in my hardware, so I've added the option to ignore them at runtime. However, I do realise not all kilns are the same and, so I have made the "Ignore" option something you have to actively enable.

# Abort on various errors

I've made it shut down the kiln immediately if connection to the thermocouple is lost or an unknown error is detected. I'm not sure if people occasionally see these because of glitches, so perhaps this is too draconic; If so, I can move remove it in favour of the next option:

I've also added an error threshold system. Even though I'm going to be ignoring VCC/GND errors, I thought others might appreciate this for safety reasons. It will shut down the kiln if you unexpectedly start seeing **lots** of errors in a short period. However, it should still allow the occasional glitch through (I think you mentioned you have this sometimes?).

Finally, I've made it refuse to start the kiln on any sort of thermocouple error no matter what - my assumption is if you're seeing the VCC/GND errors at room temperature, they're _really_ a problem!

# Moving average temperature and configurable sample count

This attempts to solve two issues:
* At ~920C the readings get a bit noisier (I see spikes of +- 4C reasonably frequently), which causes the PID loop problems. 
* Occasionally, the temperature and oven threads get slightly out of sync and the existing read-5-samples-and-update method  means the oven sometimes reads a temperature lagging 2 more seconds behind than normal. Not a huge problem really, but I thought I'd mention it since I happened to spot it.

So: I've switched it to a moving average updating the oven temperature every time a new sample is read, so the oven always has access to the latest (averaged) temperature.

I also added a configuration option to allow the number of averaged samples to be changed. I set the default to 5 as it is now, but I found I needed a value of about 40 to smooth out the random spikes at high temperatures.

jbruce12000 reviewed (commented):
I've never tried a number as high as 40 for getting the average across one duty cycle. Have you tested what the max for that is?

adq commented:
Good question!

Looking at the datasheet, it says "typical" conversion time 70ms, max 100ms, which implies its doing max 10 conversions/sec, but there can be a variance in the exact timing of a conversion.

It seems the SPI interface is not tied to the conversion timing: I'm totally able to max a (bitbanged pi zero) CPU reading from it at ~500 reads/sec with no errors whatsoever (as long as I am below 900C obviously in my kiln's case).

There must be a buffer between SPI and the raw ADC hardware to insulate it from reading half-completed conversions (given I don't get any sort of weird readings or errors even at 500/sec)

Since the temperature conversions can have an unknown variance in their exact timing, I think a >1 multiple of 10 reads/sec is sensible to prevent any weird aliasing issues resulting from the conversion variance. 40 was actually just a "gut feeling" guess that seemed to work on my system.

Giterd commented:
I would be happy to see this implemented.  It worked a treat for me and solve my short-to-ground problems.

adq commented:
Now that I own an adafruit MAX31856 board, I've been doing a spot more work here:

- Configurable line frequency filter setting (50Hz vs 60Hz) for max31856.
- Exposed the max31856 avgsel parameter since it seems you can get the chip to do the anti-noise multi-temperature-sample averaging all by itself now! This is really just exposed for easier playing.
- I modified the HEAT/COOL cycle slightly: previously it always enabled/disabled the GPIO every cycle even if there was supposed to be no heating/cooling during that cycle. Now it doesn't. Not sure how much difference it will make, but it seems more correct.

All still very experimental - my kiln is active right now so not tested yet.

Incidentally, I tested out my original Orton kiln controller again. I think this software may now be beating it (and not just from a GUI POV) - the PID loop seems a lot more stable!

adq commented:
Spot more tweaking/testing:

- Removed 'avgsel' again and left it at the default of 0 -- we've already got our own averaging code, so doing it on the chip doesn't add anything useful.
- I enabled and implemented open circuit detection for the max31856 (previously it was disabled by the default config).

Interestingly, the max31856 doesn't seem to support the shorttoground/shorttovcc error reporting stuff at all. Not yet fired it up to 900C, but will try that out later.

Giterd commented:
@adq ,Thanks for continuing with this effort.  I am still running your 'temp' version, but looking forward to reverting to this master once it has been merged.

adq commented:
@Giterd thanks! Its working really well for me now.

BTW, if you ever wanted to try my later stuff, my `all` branch at https://github.com/adq/kiln-controller/tree/all has all my changes on top of the latest from Jason's upstream.

jbruce12000 reviewed (commented):
busy this weekend so I don't have time to test. I really like the moving average idea. I want to test this and make sure it works ok.

jbruce12000 commented:
please don't stack features in the future. Separate them out so they are easier to assess, test, and review.
